README.md
# KTM for 1.14.x Hermes Proxy
Based on https://github.com/dfherr/ThreatClassic2 https://github.com/dfherr/LibThreatClassic2

## License

[MIT](license/ThreatClassic2)

Copyright (c) 2019 Dennis-Florian Herr


CHANGELOG.md
# Changelog

## v1.3

* fixed onyxia knock away (on vmangos, threat is reduced even if spell misses)

## v1.2

* Added UI for setting master target and clearing raid threat. Also added
    the same restrictions on who can send those messages as the 1.12 KTM.
    These actions can now also be set to keybinds in the Blizzard "Key
    Bindings" UI

    The previous /ktm mt command now also supports supplying a name so a
    master target can be set at any time. (e.g. /ktm mt Nefarian)

* Better support for Thunderfury threat due to the limitations of
    detecting aura refreshes using Hermes/Vmagnos and a difference in
    behavior of the combat log compared to classic

## v1.1

* Disabled rested check so players can test on the training dummys on the Everlook server

## v1.0.6

* feral druids now tracked as melee
* removed 125% modifier on execute (vmangos is incorrectly missing the scaling on execute)

## v1.0.5

* fixed bug in 1.0.4 release

## v1.0.4

* ThreatClassic2 was assuming all threat was in melee range. Now we assume they are ranged if they are not a warrior or rogue.
* fixed max threat broken when a master target is set
* fixed taunt gaining threat when spell fails

## v1.0.3

* fixed bug preventing addon from working for druids


locale/enUS.lua
local TC2, C, L, _ = unpack(select(2, ...))

-----------------------------
--	enUS and enGB clients
-----------------------------
-- main frame
L.gui_threat			= "KTMClassic"
L.gui_config			= "Open Config"

-- config frame
L.reset					= "Reset to Defaults"

-- messages
L.message_welcome		= "Type /ktm for options."
L.message_leader		= "You must be the group leader or have assist to do this."
L.message_outdated		= "New version of LibThreatClassic2 available! Please download the latest version. Latest version is always available in ThreatClassic2 and shortly after in DBM, Details and some other addons."
L.message_incompatible	= "TC2/LibThreatClassic2 is out of date and will not be compatible for this group. Please upgrade immediately."

-- version check
L.version				= "Version Check"
L.version_check			= "Check Outdated"
L.version_check_all		= "Check All"
L.version_notify		= "Notify Outdated"
L.version_list			= "Version Check: All Clients"
L.version_list_outdated	= "Version Check: Outdated Clients"
L.version_divider		= "--------------------------------"
L.version_incompatible	= "Not compatible!"

L.general				= "General"
L.general_welcome		= "Display welcome message when logging in."
L.general_rawPercent    = "Show raw threat percentage (aggro at 110% melee / 130% range)"
L.general_updateFreq    = "Time (in seconds) between updates."
L.general_test			= "Enable test mode."
L.general_minimap		= "Toggle minimap icon."
L.general_ignorePets	= "Ignore threat from player pets."

L.visibility			= "Visibility"
L.visibility_hideOOC	= "Hide frame when out of combat."
L.visibility_hideSolo	= "Hide frame when not in a group."
L.visibility_hideInPvP	= "Hide frame when in battlegrounds."
L.visibility_hideOpenWorld	= "Hide frame when not in an instance (also hides world bosses!)."
L.visibility_hideAlways	= "Hide frame always. /tc2 toggle"

L.nameplates			= "Nameplates"
L.nameplates_enable		= "Enable threat coloring (only for Blizzard nameplates)."
L.nameplates_invert		= "Invert threat coloring (intended for tanks)."
L.nameplates_colors		= "Threat Colors"

L.profiles              = "Profiles"

L.color					= "Colors"
L.color_good			= "Good"
L.color_neutral			= "Neutral"
L.color_bad				= "Bad"

L.appearance			= "Appearance"

L.frame					= "Frame"
L.frame_header			= "Header"
L.frame_bg				= "Background"
L.frame_test			= "Test Mode"
L.frame_strata			= "Strata"
L.frame_scale			= "Scale"
L.frame_lock			= "Lock"
L.frame_headerShow		= "Show Header"
L.frame_headerColor		= "Header Color"
L.frame_position		= "Position"
L.frame_width			= "Width"
L.frame_height			= "Height"
L.frame_xOffset         = "X Offset"
L.frame_yOffset         = "Y Offset"

L.bar					= "Bars"
L.bar_count				= "Maximum Count"
L.bar_descend			= "Reverse Growth"
L.bar_height			= "Height"
L.bar_padding			= "Padding"
L.bar_marker			= "Player Threat in Red"
L.bar_texture			= "Texture"
L.bar_classColor		= "Use Class Color"
L.bar_defaultColor		= "Custom Color"
L.bar_alpha				= "Bar Alpha (default colors)"
L.bar_colorMod			= "Color Modifier"

L.customBarColors 			        = "Custom Bar Colors"
L.customBarColorsPlayer_enabled	    = "Enable Custom Player Color"
L.customBarColorsActiveTank_enabled	= "Enable Custom Active Tank Color"
L.customBarColorsOtherUnit_enabled 	= "Enable Custom Other Unit Color"
L.customBarColorsPlayer_color 	    = "Player Color"
L.customBarColorsActiveTank_color 	= "Active Tank Color"
L.customBarColorsOtherUnit_color 	= "Other Unit Color"

L.font					= "Font"
L.font_name			    = "Name"
L.font_size				= "Size"
L.font_style			= "Style"
L.font_shadow			= "Dropshadow"

L.warnings				= "Warnings"
L.warnings_flash		= "Enable screen flashing."
L.warnings_soundHeader	= "Sound warning settings."
L.warnings_sound		= "Enable sounds."
L.warnings_threshold	= "Warning threshold (100% = aggro)."
L.warnings_minThreatAmount = "Minimum threat to trigger warnings."

L.warnings_soundFile	= "Sound file"
L.warnings_soundChannel	= "Sound channel"

L.soundChannel_master = "Master"
L.soundChannel_sfx = "SFX"
L.soundChannel_ambience = "Ambience"
L.soundChannel_music = "Music"


Libs/LibThreatClassic2/NPCModules/Onyxia/Onyxia.lua
if not _G.THREATLIB_LOAD_MODULES then return end -- only load if LibThreatClassic2.lua allows it
if not LibStub then return end
local ThreatLib, MINOR = LibStub("LibThreatClassic2", true)
if not ThreatLib then return end

local ONYXIA_ID = 10184
local FIREBALL_ID = 18392
local KNOCK_AWAY_ID = 19633

local phase3Text = "It seems you'll need another lesson, mortals!"

ThreatLib:GetModule("NPCCore-r"..MINOR):RegisterModule(ONYXIA_ID, function(Onyxia)
	Onyxia:RegisterTranslation("enUS", function() return {
		[phase3Text] = phase3Text,
	} end)

	Onyxia:RegisterTranslation("deDE", function() return {
		[phase3Text] = nil, -- STRING NEEDED
	} end)

	Onyxia:RegisterTranslation("esES", function() return {
		[phase3Text] = nil, -- STRING NEEDED
	} end)

	Onyxia:RegisterTranslation("esMX", function() return {
		[phase3Text] = nil, -- STRING NEEDED
	} end)

	Onyxia:RegisterTranslation("frFR", function() return {
		[phase3Text] = nil, -- STRING NEEDED
	} end)

	Onyxia:RegisterTranslation("itIT", function() return {
		[phase3Text] = nil, -- STRING NEEDED
	} end)

	Onyxia:RegisterTranslation("koKR", function() return {
		[phase3Text] = nil, -- STRING NEEDED
	} end)

	Onyxia:RegisterTranslation("ptBR", function() return {
		[phase3Text] = nil, -- STRING NEEDED
	} end)

	Onyxia:RegisterTranslation("ruRU", function() return {
		[phase3Text] = nil, -- STRING NEEDED
	} end)

	Onyxia:RegisterTranslation("zhCN", function() return {
		[phase3Text] = nil, -- STRING NEEDED
	} end)

	Onyxia:RegisterTranslation("zhTW", function() return {
		[phase3Text] = nil, -- STRING NEEDED
	} end)

	local phase3 = Onyxia:GetTranslation(phase3Text)

	function Onyxia:Init()
        self:RegisterCombatant(ONYXIA_ID, true)
		self:RegisterSpellHandler("SPELL_CAST_SUCCESS", ONYXIA_ID, FIREBALL_ID, self.Fireball)
		self:RegisterSpellHandler("SPELL_CAST_SUCCESS", ONYXIA_ID, KNOCK_AWAY_ID, self.KnockAway)
		self:RegisterChatEvent("yell", phase3, self.Phase3Start)
	end

    function Onyxia:Fireball(sourceGUID, unitId)
        self:ModifyThreat(sourceGUID, unitId, 0, 0)
	end

	function Onyxia:KnockAway(sourceGUID, unitId)
        self:ModifyThreat(sourceGUID, unitId, 0.75, 0)
	end

	function Onyxia:Phase3Start(sourceGUID, unitId)
        self:WipeRaidThreatOnMob(ONYXIA_ID)
	end
end)


Libs/LibThreatClassic2/ClassModules/Classic/Druid.lua
if not _G.THREATLIB_LOAD_MODULES then return end -- only load if LibThreatClassic2.lua allows it
if not LibStub then return end
local ThreatLib, MINOR = LibStub("LibThreatClassic2", true)
if not ThreatLib then return end

if select(2, _G.UnitClass("player")) ~= "DRUID" then return end

local _G = _G
local pairs = _G.pairs
local select = _G.select
local GetTalentInfo = _G.GetTalentInfo
local GetShapeshiftForm = _G.GetShapeshiftForm
local UnitGUID = _G.UnitGUID

local Druid = ThreatLib:GetOrCreateModule("Player-r"..MINOR)

local faerieFireFactor = 108 / 54	-- NEED MORE INFO

local threatValues = {
	["cower"] = {
		[8998] = 240,
		[9000] = 390,
		[9892] = 600
	},
	["demoralizingRoar"] = {
		[99] = 9,
		[1735] = 15,
		[9490] = 20,
		[9747] = 30,
		[9898] = 39
	},
	["faerieFire"] = {
		-- normal
		[770] = faerieFireFactor * 18,
		[778] = faerieFireFactor * 30,
		[9749] = faerieFireFactor * 42,
		[9907] = 108,
		-- feral
		[16857] = faerieFireFactor * 18,
		[17390] = faerieFireFactor * 30,
		[17391] = faerieFireFactor * 42,
		[17392] = 108,
	},
}

local SCHOOL_MASK_NATURE = _G.SCHOOL_MASK_NATURE or 0x08
local SCHOOL_MASK_ARCANE = _G.SCHOOL_MASK_ARCANE or 0x40

local itemSets = {}
local swipeIDs = {779, 780, 769, 9754, 9908}
local maulIDs = {6807, 6808, 6809, 8972, 9745, 9880, 9881}
local tranquilityIDs = {740, 8918, 9862, 9863}

function Druid:ClassInit()

	-- Growl
	self.CastSuccessHandlers[6795] = self.Growl

	for k, v in pairs(threatValues.faerieFire) do
		self.CastSuccessHandlers[k] = self.FaerieFire
		self.CastMissHandlers[k] = self.FaerieFireMiss
	end
	for k, v in pairs(threatValues.cower) do
		self.CastSuccessHandlers[k] = self.Cower
		self.CastMissHandlers[k] = self.CowerMiss
	end

	for k, v in pairs(threatValues.demoralizingRoar) do
		self.MobDebuffHandlers[k] = self.DemoralizingRoar
	end

	-- Subtlety for all Arcane or Nature Damage, as well as heals.
	self.schoolThreatMods[SCHOOL_MASK_NATURE] = self.Subtlety
	self.schoolThreatMods[SCHOOL_MASK_ARCANE] = self.Subtlety

	for i = 1, #swipeIDs do
		self.AbilityHandlers[swipeIDs[i]] = self.Swipe
	end
	swipeIDs = nil

	for i = 1, #maulIDs do
		self.AbilityHandlers[maulIDs[i]] = self.Maul
	end
	maulIDs = nil

	for i = 1, #tranquilityIDs do
		self.AbilityHandlers[tranquilityIDs[i]] = self.Tranquility
	end
	tranquilityIDs = nil

	self.itemSets = itemSets
end

function Druid:ClassEnable()
	self:GetStanceThreatMod()
	self:RegisterEvent("UPDATE_SHAPESHIFT_FORM", "GetStanceThreatMod")
end

function Druid:ClassDisable()
	self:UnregisterEvent("UPDATE_SHAPESHIFT_FORM")
end

function Druid:ScanTalents()
	if ThreatLib.Classic then
		self.feralinstinctMod = 0.03 * select(5, GetTalentInfo(2, 3))
		self.subtletyMod = 1 - 0.04 * select(5, GetTalentInfo(3, 8))
		self.tranqMod = 1 - 0.5 * select(5, GetTalentInfo(3, 13))
	else
		self.feralinstinctMod = 0 -- for when testing in retail
		self.subtletyMod = 1 -- for when testing in retail
		self.tranqMod = 1 -- for when testing in retail
	end
end

-- get form passive threat modifier
function Druid:GetStanceThreatMod()
	local form = GetShapeshiftForm()
	self.isTanking = false
	if form == 1 then
		self.passiveThreatModifiers = 1.3 + self.feralinstinctMod
		self.isTanking = true
	elseif form == 2 or form == 3 then
		-- if aquatic form is not learnt, druid cat form is 2 and travel form 3
		-- else aquatic form is 2 and cat form 3.
		-- should be fine, druids don't dps in travel or aquatic form xD
		self.passiveThreatModifiers = 0.71
	else
		self.passiveThreatModifiers = 1
	end
	self.totalThreatMods = nil -- Needed to recalc total mods
end
local pendingTauntTarget = nil
local pendingTauntOffset = nil
function Druid:Growl(spellID, target)
	local targetThreat = ThreatLib:GetThreat(UnitGUID("targettarget"), target)
	local myThreat = ThreatLib:GetThreat(UnitGUID("player"), target)
	if targetThreat > 0 and targetThreat > myThreat then
		pendingTauntTarget = target
		pendingTauntOffset = targetThreat-myThreat
	elseif targetThreat == 0 then
		local maxThreat = ThreatLib:GetMaxThreatOnTarget(target)
		pendingTauntTarget = target
		pendingTauntOffset = maxThreat-myThreat
	end
	self.nextEventHook = self.GrowlNextHook
end

function Druid:GrowlNextHook(timestamp, subEvent, hideCaster, sourceGUID, sourceName, sourceFlags, sourceRaidFlags, destGUID, destName, destFlags, destRaidFlags, spellID)
	if pendingTauntTarget and (subEvent ~= "SPELL_MISSED" and spellID ~= 6795) then
		self:AddTargetThreat(pendingTauntTarget, pendingTauntOffset)
		ThreatLib:PublishThreat()
	end
	pendingTauntTarget = nil
	pendingTauntOffset = nil
end

function Druid:FaerieFire(spellID, target)
	self:AddTargetThreat(target, threatValues.faerieFire[spellID] * self:threatMods())
end

function Druid:FaerieFireMiss(spellID, target)
	self:AddTargetThreat(target, -(threatValues.faerieFire[spellID] * self:threatMods()))
end

function Druid:Maul(amount)
	return amount * 1.75
end

function Druid:Swipe(amount)
	return amount * 1.75
end

function Druid:Cower(spellID, target)
	self:AddTargetThreat(target, threatValues.cower[spellID] * self:threatMods() * -1)
end

function Druid:CowerMiss(spellID, target)
	self:AddTargetThreat(target, threatValues.cower[spellID] * self:threatMods())
end

function Druid:DemoralizingRoar(spellID, target)
	self:AddTargetThreat(target, threatValues.demoralizingRoar[spellID] * self:threatMods())
end

function Druid:Subtlety(amount)
	return amount * self.subtletyMod
end

function Druid:Tranquility(amount)
	return amount * self.tranqMod
end


Libs/LibThreatClassic2/ClassModules/Classic/Warrior.lua
if not _G.THREATLIB_LOAD_MODULES then return end -- only load if LibThreatClassic2.lua allows it
if not LibStub then return end
local ThreatLib, MINOR = LibStub("LibThreatClassic2", true)
if not ThreatLib then return end

if select(2, _G.UnitClass("player")) ~= "WARRIOR" then return end

local _G = _G
local select = _G.select
local GetTalentInfo = _G.GetTalentInfo
local GetShapeshiftForm = _G.GetShapeshiftForm
local GetSpellInfo = _G.GetSpellInfo
local pairs, ipairs = _G.pairs, _G.ipairs
local GetTime = _G.GetTime
local UnitDebuff = _G.UnitDebuff

local Warrior = ThreatLib:GetOrCreateModule("Player-r"..MINOR)

-- https://github.com/magey/classic-warrior/wiki/Threat-Mechanics
-- maxRankThreatValue / maxRankLevelAvailability = factor
-- lowerRankThreatValue = factor * rankN

local sunderFactor 			= 261 / 58	-- ok-ish (between 260 - 270)
local shieldBashFactor 		= 187 / 52
local revengeFactor 		= 355 / 60	-- maybe look into r5
local heroicStrikeFactor 	= 220 / 70	-- NEED MORE INFO
local shieldSlamFactor 		= 250 / 60
local cleaveFactor 			= 100 / 60	-- NEED MORE INFO
local hamstringFactor 		= 145 / 54	-- ok-ish (between 142 - 148)
local mockingBlowFactor 	= 250 / 56	-- NEED MORE INFO
local battleShoutFactor 	= 1
local demoShoutFactor 		= 43 / 54
local thunderClapFactor		= 130 / 58

local threatValues = {
	sunder = {
		[7386] = sunderFactor * 10,
		[7405] = sunderFactor * 22,
		[8380] = sunderFactor * 34,
		[11596] = sunderFactor * 46,
		[11597] = 261
	},
	shieldBash = {
		[72] = shieldBashFactor * 12,
		[1671] = shieldBashFactor * 32,
		[1672] = 187
	},
	revenge = {
		[6572] = revengeFactor * 14,
		[6574] = revengeFactor * 24,
		[7379] = revengeFactor * 34,
		[11600] = revengeFactor * 44,
		[11601] = 315,
		[25288] = 355
	},
	heroicStrike = { -- needs work
		[78] = 20,
		[284] = 39,
		[285] = 59,
		[1608] = 78,
		[11564] = 98,
		[11565] = 118,
		[11566] = 137,
		[11567] = 145,
		[25286] = 175
	},
	shieldSlam = {
		[23922] = shieldSlamFactor * 40,
		[23923] = shieldSlamFactor * 48,
		[23924] = shieldSlamFactor * 54,
		[23925] = 250
	},
	cleave = {
		[845] = 10,
		[7369] = 40,
		[11608] = 60,
		[11609] = 70,
		[20569] = 100
	},
	hamstring = {
		[1715] = hamstringFactor * 8,
		[7372] = hamstringFactor * 32,
		[7373] = 145,
	},
	--[[
	mockingBlow = {
		[694] = mockingBlowFactor * 16,
		[7400] = mockingBlowFactor * 26,
		[7402] = mockingBlowFactor * 36,
		[20559] = mockingBlowFactor * 46,
		[20560] = 250
	},
	--]]
	battleShout = {
		[6673] = battleShoutFactor * 1,
		[5242] = battleShoutFactor * 12,
		[6192] = battleShoutFactor * 22,
		[11549] = battleShoutFactor * 32,
		[11550] = battleShoutFactor * 42,
		[11551] = battleShoutFactor * 52,
		[25289] = 60
	},
	demoShout = {
		[1160] = demoShoutFactor * 14,
		[6190] = demoShoutFactor * 24,
		[11554] = demoShoutFactor * 34,
		[11555] = demoShoutFactor * 44,
		[11556] = 43
	},
	thunderclap = {
		[6343] = thunderClapFactor * 6,
		[8198] = thunderClapFactor * 18,
		[8204] = thunderClapFactor * 28,
		[8205] = thunderClapFactor * 38,
		[11580] = thunderClapFactor * 48,
		[11581] = 130
	},
	execute = {
		[5308] = true,
		[20658] = true,
		[20660] = true,
		[20661] = true,
		[20662] = true
	},
	disarm = {
		[676] = 104
	}
}

local function init(self, t, f)
	local funcSuccess = function(self, spellID, target)
		self:AddTargetThreat(target, f(self, spellID))
	end
	local funcMiss = function(self, spellID, target)
		self:AddTargetThreat(target, -f(self, spellID))
	end
	for k, v in pairs(t) do
		self.CastSuccessHandlers[k] = funcSuccess
		self.CastMissHandlers[k] = funcMiss
	end
end

function Warrior:ClassInit()
	-- -- Taunt. Assumption that all successful taunts apply a debuff to the mob
	self.CastSuccessHandlers[355] = self.Taunt

	init(self, threatValues.heroicStrike, self.HeroicStrike)
	init(self, threatValues.cleave, self.Cleave)
	init(self, threatValues.shieldBash, self.ShieldBash)
	init(self, threatValues.shieldSlam, self.ShieldSlam)
	init(self, threatValues.revenge, self.Revenge)
	-- init(self, threatValues.mockingBlow, self.MockingBlow)
	init(self, threatValues.hamstring, self.Hamstring)
	init(self, threatValues.thunderclap, self.Thunderclap)
	init(self, threatValues.disarm, self.Disarm)
	init(self, threatValues.sunder, self.SunderArmor)

	-- Ability damage modifiers
	for k, v in pairs(threatValues.execute) do
		self.AbilityHandlers[k] = self.Execute
	end

	-- Shouts
	-- Battle Shout
	local bShout = function(self, spellID, target)
		self:AddThreat(threatValues.battleShout[spellID] * self:threatMods())
	end
	for k, v in pairs(threatValues.battleShout) do
		self.CastHandlers[k] = bShout
	end

	-- Demoralizing Shout
	local demoShoutFunc = function(self, spellID, target)
		self:AddTargetThreat(target, threatValues.demoShout[spellID] * self:threatMods())
	end
	for k, v in pairs(threatValues.demoShout) do
		self.MobDebuffHandlers[k] = demoShoutFunc
	end

	-- Set names don't need to be localized.
	self.itemSets = {
		["Might"] = { 16866, 16867, 16868, 16862, 16864, 16861, 16865, 16863 }
	}
end

function Warrior:ClassEnable()
	self:GetStanceThreatMod()
	self:RegisterEvent("UPDATE_SHAPESHIFT_FORM", "GetStanceThreatMod" )
end

function Warrior:ClassDisable()
	self:UnregisterEvent("UPDATE_SHAPESHIFT_FORM")
end

function Warrior:ScanTalents()
	-- Defiance
	if ThreatLib.Classic then
		local rank = _G.select(5, GetTalentInfo(3, 9))
		self.defianceMod = 1 + (0.03 * rank)
	else
		self.defianceMod = 1 -- for when testing in retail
	end
end

function Warrior:GetStanceThreatMod()
	self.isTanking = false
	if GetShapeshiftForm() == 2 then
		self.passiveThreatModifiers = 1.3 * self.defianceMod
		self.isTanking = true
	elseif GetShapeshiftForm() == 3 then
		self.passiveThreatModifiers = 0.8
	else
		self.passiveThreatModifiers = 0.8
	end
	self.totalThreatMods = nil -- Needed to recalc total mods
end

function Warrior:SunderArmor(spellID)
	local sunderMod = 1
	if self:getWornSetPieces("Might") >= 8 then
		sunderMod = 1.15
	end
	local threat = threatValues.sunder[spellID]
	return threat * sunderMod * self:threatMods()
end

local pendingTauntTarget = nil
local pendingTauntOffset = nil
-- function Warrior:Taunt(spellID, target)
-- 	print("warrior taunt! " .. UnitGUID("targettarget"))
-- 	local targetThreat = ThreatLib:GetThreat(UnitGUID("targettarget"), target, "target", true)
-- 	local myThreat = ThreatLib:GetThreat(UnitGUID("player"), target, "target", true)

-- 	print("my threat: " .. myThreat .. " targettarget threat: " .. targetThreat)
	
-- 	if targetThreat > 0 and targetThreat > myThreat then
-- 		self:AddTargetThreat(target, targetThreat-myThreat)
-- 		ThreatLib:PublishThreat()
-- 	elseif targetThreat == 0 then
-- 		local maxThreat = ThreatLib:GetMaxThreatOnTarget(target)
-- 		self:AddTargetThreat(target, maxThreat-myThreat)
-- 		ThreatLib:PublishThreat()
-- 	end
-- end

function Warrior:Taunt(spellID, target)
	local targetThreat = ThreatLib:GetThreat(UnitGUID("targettarget"), target)
	local myThreat = ThreatLib:GetThreat(UnitGUID("player"), target)
	if targetThreat > 0 and targetThreat > myThreat then
		pendingTauntTarget = target
		pendingTauntOffset = targetThreat-myThreat
	elseif targetThreat == 0 then
		local maxThreat = ThreatLib:GetMaxThreatOnTarget(target)
		pendingTauntTarget = target
		pendingTauntOffset = maxThreat-myThreat
	end
	self.nextEventHook = self.TauntNextHook
end

function Warrior:TauntNextHook(timestamp, subEvent, hideCaster, sourceGUID, sourceName, sourceFlags, sourceRaidFlags, destGUID, destName, destFlags, destRaidFlags, spellID)
	if pendingTauntTarget and (subEvent ~= "SPELL_MISSED" and spellID ~= 355) then
		self:AddTargetThreat(pendingTauntTarget, pendingTauntOffset)
		ThreatLib:PublishThreat()
	end
	pendingTauntTarget = nil
	pendingTauntOffset = nil
end

function Warrior:HeroicStrike(spellID)
	return threatValues.heroicStrike[spellID] * self:threatMods()
end

function Warrior:Cleave(spellID)
	return threatValues.cleave[spellID] * self:threatMods()
end

function Warrior:ShieldBash(spellID)
	return threatValues.shieldBash[spellID] * self:threatMods()
end

function Warrior:ShieldSlam(spellID)
	return threatValues.shieldSlam[spellID] * self:threatMods()
end

function Warrior:Revenge(spellID)
	return threatValues.revenge[spellID] * self:threatMods()
end

--[[
function Warrior:MockingBlow(spellID)
	return threatValues.mockingBlow[spellID] * self:threatMods()
end
--]]

function Warrior:Hamstring(spellID)
	return threatValues.hamstring[spellID] * self:threatMods()
end

function Warrior:Thunderclap(spellID)
	return threatValues.thunderclap[spellID] * self:threatMods()
end

function Warrior:Execute(amount)
	return amount
end

function Warrior:Disarm(spellID)
	return threatValues.disarm[spellID] * self:threatMods()
end
